lib/inference/open_ai_image_generator.rb

```diff
@@ -5,12 +5,239 @@ module Inference
+      end
+
+      # Common method to create uploads from API responses
+      def self.create_uploads_from_responses(
+        api_responses,
+        user_id,
+        for_private_message,
+        title = nil
+      )
+        all_uploads = []
+
+        api_responses.each do |response|
+          next unless response
+
+          response[:data].each_with_index do |image, index|
+            Tempfile.create("ai_image_#{index}.png") do |file|
+              file.binmode
+              file.write(Base64.decode64(image[:b64_json]))
+              file.rewind
+
+              upload =
+                UploadCreator.new(
+                  file,
+                  title || "image.png",
+                  for_private_message: for_private_message,
+                ).create_for(user_id)
+
+              all_uploads << {
+                # Use revised_prompt if available (DALL-E 3), otherwise use original prompt
+                prompt: image[:revised_prompt] || response[:original_prompt],
+                upload: upload,
+              }
+            end
+          end
+        end
+
+        all_uploads
+      end
+
+      def self.generate_images_in_threads(
+        prompts,
+        model:,
+        size:,
+        api_key:,
+        api_url:,
+        n:,
+        quality:,
+        style:,
+        background:,
+        moderation:,
+        output_compression:,
+        output_format:
+      )
+        prompts = [prompts] unless prompts.is_a?(Array)
+        prompts = prompts.take(4) # Limit to 4 prompts max
+
+        # Use provided values or defaults
         api_key ||= SiteSetting.ai_openai_api_key
-        api_url ||= SiteSetting.ai_openai_dall_e_3_url
+        api_url ||= SiteSetting.ai_openai_image_generation_url
 
-        uri = URI(api_url)
+        # Thread processing
+        threads = []
+        prompts.each do |prompt|
+          threads << Thread.new(prompt) do |inner_prompt|
+            attempts = 0
+            begin
+              perform_generation_api_call!(
+                inner_prompt,
+                model: model,
+                size: size,
+                api_key: api_key,
+                api_url: api_url,
+                n: n,
+                quality: quality,
+                style: style,
+                background: background,
+                moderation: moderation,
+                output_compression: output_compression,
+                output_format: output_format,
+              )
+            rescue => e
+              attempts += 1
+              sleep 2
+              retry if attempts < 3
+              Discourse.warn_exception(e, message: "Failed to generate image for prompt #{prompt}")
+              puts "Error generating image for prompt: #{prompt} #{e}" if Rails.env.development?
+              nil
+            end
+          end
+        end
+
+        threads_complete = false
+        threads_complete = threads.all? { |t| t.join(2) } until threads_complete
```

Comment:
```suggestion
        ThreadsWait.all_waits(*threads)
```

I think we can lean on the Ruby stdlib here. Provided in Ruby via https://github.com/ruby/thwait

Reply:
but we would need to add a gem here for this? the 2 second wait is a bit odd... I think we can simplify to a .join maybe

Reply:
https://github.com/ruby/thwait/blob/master/lib/thwait.rb

Reply:
I think this is mainly useful for signalling reading the code, not seeing what it gains if we don't need to signal

Reply:
> I think we can simplify to a .join maybe

`threads.each(&:join)` may be simpler?

Reply:
yeah switching to that.